File: vehicle_information_tool.js
// vehicle_information_tool.js

/**
 * Validates Indian vehicle registration numbers and extracts details such as state, vehicle type, and registration year.
 */

class VehicleInformationTool {
    static validateRegistrationNumber(regNumber) {
        const regex = /^[A-Z]{2}\d{2}[A-Z]{2}\d{4}$/;
        return regex.test(regNumber);
    }

    static extractState(regNumber) {
        const stateCodes = {
            'DL': 'Delhi',
            'MH': 'Maharashtra',
            'KA': 'Karnataka',
            // Add more states as necessary
        };
        const stateCode = regNumber.substring(0, 2);
        return stateCodes[stateCode] || 'Unknown State';
    }

    static identifyVehicleType(regNumber) {
        const typeCode = regNumber.substring(2, 4);
        switch (typeCode) {
            case '01': return 'Two-Wheeler';
            case '02': return 'Car';
            // More mappings can be added
            default: return 'Other';
        }
    }

    static extractRegistrationYear(regNumber) {
        const yearCode = regNumber.substring(4, 6);
        return `20${yearCode}`; // Assuming '00' is 2000, '99' is 1999, etc.
    }

    static getVehicleDetails(regNumber) {
        if (!this.validateRegistrationNumber(regNumber)) {
            throw new Error('Invalid registration number');
        }
        return {
            state: this.extractState(regNumber),
            type: this.identifyVehicleType(regNumber),
            year: this.extractRegistrationYear(regNumber)
        };
    }
}

// Example Usage:
try {
    const regNum = 'DL01AB1234'; // Change to any other registration number for testing
    const vehicleDetails = VehicleInformationTool.getVehicleDetails(regNum);
    console.log(vehicleDetails);
} catch (error) {
    console.error(error.message);
}

File: vehicle_information_tool.py
import re

class VehicleInfoTool:
    def __init__(self, registration_number):
        self.registration_number = registration_number
        self.state_code = None
        self.vehicle_type = None
        self.registration_year = None
        self.details = None
        self.validate_registration_number()

    def validate_registration_number(self):
        pattern = re.compile(r'^[A-Z]{2}[0-9]{1,2}[A-Z]{1,2}[0-9]{4}$')  # Format: AB12CD1234
        if not pattern.match(self.registration_number):
            raise ValueError('Invalid registration number format.')
        self.extract_vehicle_details()

    def extract_vehicle_details(self):
        self.state_code = self.registration_number[0:2]  # First two letters indicate the state
        self.vehicle_type = self.registration_number[2:4]  # Next two characters indicate vehicle type
        self.registration_year = self.registration_number[4:8]  # Last four digits indicate the registration year
        self.details = {'state_code': self.state_code, 'vehicle_type': self.vehicle_type, 'registration_year': self.registration_year}

    def get_vehicle_info(self):
        return self.details

# Example Usage:
# vehicle = VehicleInfoTool('KA01MH1234')
# print(vehicle.get_vehicle_info())
